fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,96 @@
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+    fetch('/api/todos')
+      .then(res => res.json())
+      .then(todos => this.setState({ todos }));
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    fetch('/api/todos', { method: 'POST', body: JSON.stringify(newTodo), headers: { 'Content-Type': 'application/json' } })
+      .then(res => res.json())
+      .then(todos => this.setState(prevState => ({ todos: prevState.todos.concat(todos) })));
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     * replace the old todo in the state array.
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list should "toggle" its isCompleted status back and forth.
+     *
+     * Do not mutate the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    const copy = this.state.todos.map(todo => ({ ...todo }));
+    for (let i = 0; i < copy.length; i++) {
+      if (copy[i].todoId === todoId) {
+        copy[i].isCompleted = !copy[i].isCompleted;
+        fetch(`/api/todos/${todoId}`, {
+          method: 'PATCH',
+          headers: {
+            'Content-Type': 'application/json'
+          },
+          body: JSON.stringify(copy[i])
+        })
+          .then(response => response.json())
+          .then(data => this.setState({ todos: copy })
+          );
+      }
+    }
```

Comment:
Even though this will work and should only make one request we really don't want to be in the habit of making a `fetch` request from inside a `for` loop. Also state should be updated with the data directly from the server. The server is the single source of truth so we never want to "guess" what the server will do, we want to use the actual response from the server to update the view. Here are the steps that should happen:

```JAVASCRIPT
// Use `todoId` and this.state.todos.findIndex() to get the index of the item to update
// Get the opposite of that items isCompleted value
// Build the object to send to the server with the value from the previous step

// Make fetch request with proper settings and values
// then get response from server, parse json
// then make a copy of the todos array from state
  // Replace the correct item in the copied array with the data from server
  // setState with updated copied array
```